Baremetal multicore support

aykevl opened the pull request:
Use multiple cores of the "baremetal" riscv-qemu target. The goal is to iron out all the bugs on an easy-to-debug target before moving on to more interesting targets like RP2040. After this PR is done, the RP2040 should be relatively easy to add support for (since I already wrote most of that code).

Work in progress. This already mostly works, but there are a few things that still need to be fixed:

  - [x] ~There's a timing bug where if one goroutine pauses, and another already running goroutine pauses for a shorter time, the second may not be woken up at the earlier time.~ fixed!
  - [x] Timers (`time.NewTimer` etc) need to be implemented.
  - [x] Stack scanning (for the GC) needs to be implemented.

aykevl committed:
all: add support for multicore scheduler

This commit adds support for a scheduler that runs a scheduler on all
available cores. It is meant to be used on baremetal systems with a
fixed number of cores, such as the RP2040.

The initial implementation adds support for multicore scheduling to the
riscv-qemu target as a convenient testing target. This means that this
new multicore scheduler is tested in CI, including a bunch of standard
library tests (`make tinygo-test-baremetal`). This should ensure the new
scheduler is reasonably well tested before trying to use it on
harder-to-debug targets like the RP2040.

aykevl commented:
Rebased and fixed one of the 3 issues listed above.

aykevl commented:
Implemented support for timers.

aykevl committed:
sync: fix TestMutexConcurrent test

Accessing the same variable from multiple goroutines is unsafe, and will
fail with parallelism. A lightweight way to avoid issues is by using
atomic variables.

aykevl commented:
Implemented support for stack scanning. `make tinygo-test-baremetal` now passes. I will clean up the code and push it later.

aykevl committed:
runtime: add exportedFuncPtr

This function isn't used yet, but will be used for the "cores" scheduler
for RISC-V.

aykevl committed:
runtime/interrupt: add Checkpoint type

This type can be used to jump back to a previous position in a program
from inside an interrupt. This is useful for baremetal systems that
implement wfi but not wfe, and therefore have no easy (race-free) way to
wait until a flag gets changed inside an interrupt. This is an issue on
RISC-V, where this is racy (the interrupt might happen after the check
but before the wfi instruction):

    configureInterrupt()
    for flag.Load() != 0 {
        riscv.Asm("wfi")
    }

deadprogram commented:
@aykevl looks like legit test failure here:
https://github.com/tinygo-org/tinygo/actions/runs/15566983342/job/43835974809#step:22:43

aykevl commented:
I haven't finished pushing all the changes, so this is expected. I'll comment when I'm done.

aykevl committed:
runtime: refactor obtaining the system stack

The system stack is only needed when we're not on it. So we can directly
call task.SystemStack() without problems.

This also saves a tiny bit of binary size.

aykevl commented:
Pushed all my changes. I did a bunch of local testing and things seem good, I hope CI agrees and I didn't break something.

(With careful review of my local changes I also discovered a few bugs that weren't caught during normal testing, so this was actually very necessary).

aykevl commented:
All tests have passed! 🎉 🎉 🎉

deadprogram commented:
This PR is ready for review cc @eliasnaur @ysoldak @ardnew @kenbell @soypat anyone else that I missed who is into this sort of thing :thread: :thread: :thread:

deadprogram reviewed (approved):
> This means that this
new multicore scheduler is tested in CI, including a bunch of standard
library tests (`make tinygo-test-baremetal`)

That is completely awesome! Very good work on this @aykevl

aykevl commented:
I seem to have actually working multicore support for the RP2040 now. Will make a PR once I've cleaned up the code.

deadprogram commented:
Last call on feedback on this PR before merge! :bell:

deadprogram commented:
Great work @aykevl now merging!

ardnew commented:
Sorry, just saw this — congrats on the effort! This ~is~ **_was_** one of those features causing friction for new users, regardless if the friction was real or imagined.
